src/ncp/ncp_uart.cpp

```diff
@@ -220,10 +220,23 @@ void NcpUart::EncodeAndSendToUart(void)
 
     if (len > 0)
     {
+#if OPENTHREAD_UART_LOCK
+        if (otPlatUartLockAcquire() != OT_ERROR_NONE)
+        {
+            assert(false);
+        }
+#endif
```

Comment:
How about having the platform/driver code handling the lock from `otPlatUartSend()` call instead of having in NCP and CLI code?

The `otPlatUartSend()` implementation for a specific platform can then choose to include a lock or not.

Reply:
This PR is mainly for logging. How logging is handled depends on how OpenThread runs, NCP, CLI or other mode. In NCP mode, we cannot simply print logging into UART. If no CLI or NCP is running, we don't need such lock. So I think the lock should be added to CLI module in CLI mode as I suggested in my last comment. Thoughts?